policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,712 @@
+            // watches will complete.  No other parts of the index need to be
+            // updated.
+            if ns.get_mut().pods.by_name.remove(&name).is_some() && ns.get().is_empty() {
+                ns.remove();
+            }
+        }
+    }
+
+    // Since apply only reindexes a single pod at a time, there's no need to
+    // handle resets specially.
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::Server> for Index {
+    fn apply(&mut self, srv: k8s::policy::Server) {
+        let namespace = srv.namespace().expect("server must be namespaced");
+        let name = srv.name();
+        let server = server::Server::from_resource(srv, &self.cluster_info);
+        self.namespaces
+            .get_or_default_with_reindex(namespace, |ns| ns.policy.update_server(name, server))
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        self.namespaces
+            .get_with_reindex(namespace, |ns| ns.policy.servers.remove(&name).is_some())
+    }
+
+    fn reset(&mut self, srvs: Vec<k8s::policy::Server>, deleted: HashMap<String, HashSet<String>>) {
+        #[derive(Default)]
+        struct Ns {
+            added: Vec<(String, server::Server)>,
+            removed: HashSet<String>,
+        }
+
+        // Aggregate all of the updates by namespace so that we only reindex
+        // once per namespace.
+        let mut updates_by_ns = HashMap::<String, Ns>::default();
+        for srv in srvs.into_iter() {
+            let namespace = srv.namespace().expect("server must be namespaced");
+            let name = srv.name();
+            let server = server::Server::from_resource(srv, &self.cluster_info);
+            updates_by_ns
+                .entry(namespace)
+                .or_default()
+                .added
+                .push((name, server));
+        }
+        for (ns, names) in deleted.into_iter() {
+            updates_by_ns.entry(ns).or_default().removed = names;
+        }
+
+        for (namespace, Ns { added, removed }) in updates_by_ns.into_iter() {
+            if added.is_empty() {
+                self.namespaces.get_with_reindex(namespace, |ns| {
+                    let changed = !removed.is_empty();
+                    for name in removed.into_iter() {
+                        ns.policy.servers.remove(&name);
+                    }
+                    changed
+                });
+            } else {
+                self.namespaces
+                    .get_or_default_with_reindex(namespace, |ns| {
+                        let mut changed = !removed.is_empty();
+                        for name in removed.into_iter() {
+                            ns.policy.servers.remove(&name);
+                        }
+                        for (name, server) in added.into_iter() {
+                            changed = ns.policy.update_server(name, server) || changed;
+                        }
+                        changed
+                    });
+            }
```

Comment:
i don't totally get why the if/else is necessary here; AFAICT, the `for` loop over `added.into_iter()` will just not execute if `added` is empty, and the code would be equivalent to the code in the `if added.is_empty()` block? or am i missing something here?

Reply:
oh, i see that `get_with_reindex` also removes the namespace if it is empty. hmm. i don't _love_ the naming `get_with_reindex`/`get_or_default_with_reindex`, because it doesn't really indicate that `get_with_reindex` could also delete an empty namespace...but i'm not sure how to make that more obvious. maybe we just have to add a comment here explaining why we need to call one of those two functions, instead?